File: tgbot/src/keyboards.ts
import { b64urlEncode, CB } from "./constants";
import { monthName, statusLabel, t } from "./i18n";
import type {
  InlineKeyboardButton,
  InlineKeyboardMarkup,
  Lang,
  Profile,
  ReportStatus,
  Satellite,
  WizardState,
} from "./types";

function btn(text: string, callback_data: string): InlineKeyboardButton {
  return { text, callback_data };
}

export function satelliteKeyboard(
  sats: Satellite[],
  page: number,
  totalPages: number,
  lang: Lang
): InlineKeyboardMarkup {
  const rows: InlineKeyboardButton[][] = [];
  const start = page * 8;
  const slice = sats.slice(start, start + 8);
  for (const s of slice) {
    const label = s.display_name ?? s.name;
    rows.push([btn(label, `${CB.satPick}:${b64urlEncode(s.name)}`)]);
  }
  const nav: InlineKeyboardButton[] = [];
  if (page > 0) nav.push(btn("◀️", `${CB.satPage}:${page - 1}`));
  if (page < totalPages - 1) nav.push(btn("▶️", `${CB.satPage}:${page + 1}`));
  if (nav.length) rows.push(nav);
  rows.push([btn(`${t(lang, "btnSearch")}`, CB.satSearch), btn(t(lang, "btnCancel"), CB.cancel)]);
  return { inline_keyboard: rows };
}

export function statusKeyboard(
  statuses: ReportStatus[],
  lang: Lang,
  selected?: string
): InlineKeyboardMarkup {
  const rows: InlineKeyboardButton[][] = [];
  for (const s of statuses) {
    const label = statusLabel(lang, s.value, s.label) + (s.value === selected ? " ✓" : "");
    rows.push([btn(label, `${CB.statusPick}:${b64urlEncode(s.value)}`)]);
  }
  rows.push([btn(t(lang, "btnBack"), CB.back), btn(t(lang, "btnCancel"), CB.cancel)]);
  return { inline_keyboard: rows };
}

export function yearKeyboard(
  currentYear: number,
  page: number,
  totalPages: number,
  selected: number | undefined,
  lang: Lang
): InlineKeyboardMarkup {
  const years: number[] = [];
  for (let y = currentYear; y >= 2001; y--) years.push(y);
  const rows: InlineKeyboardButton[][] = [];
  const start = page * 10;
  const slice = years.slice(start, start + 10);
  let row: InlineKeyboardButton[] = [];
  for (const y of slice) {
    row.push(btn(`${y}${y === selected ? " ✓" : ""}`, `${CB.yearPick}:${y}`));
    if (row.length === 5) {
      rows.push(row);
      row = [];
    }
  }
  if (row.length) rows.push(row);
  const nav: InlineKeyboardButton[] = [];
  if (page > 0) nav.push(btn("◀️", `${CB.yearPage}:${page - 1}`));
  if (page < totalPages - 1) nav.push(btn("▶️", `${CB.yearPage}:${page + 1}`));
  if (nav.length) rows.push(nav);
  rows.push([
    btn(t(lang, "btnBack"), CB.back),
    btn(t(lang, "btnNext"), CB.next),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function monthKeyboard(selected: number | undefined, lang: Lang): InlineKeyboardMarkup {
  const rows: InlineKeyboardButton[][] = [];
  let row: InlineKeyboardButton[] = [];
  for (let m = 1; m <= 12; m++) {
    row.push(btn(`${monthName(lang, m)}${m === selected ? " ✓" : ""}`, `${CB.monthPick}:${m}`));
    if (row.length === 4) {
      rows.push(row);
      row = [];
    }
  }
  if (row.length) rows.push(row);
  rows.push([
    btn(t(lang, "btnBack"), CB.back),
    btn(t(lang, "btnNext"), CB.next),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function dayKeyboard(
  year: number,
  month: number,
  selected: number | undefined,
  lang: Lang
): InlineKeyboardMarkup {
  const days = new Date(Date.UTC(year, month, 0)).getUTCDate();
  const rows: InlineKeyboardButton[][] = [];
  let row: InlineKeyboardButton[] = [];
  for (let d = 1; d <= days; d++) {
    row.push(btn(`${d}${d === selected ? " ✓" : ""}`, `${CB.dayPick}:${d}`));
    if (row.length === 7) {
      rows.push(row);
      row = [];
    }
  }
  if (row.length) rows.push(row);
  rows.push([
    btn(t(lang, "btnBack"), CB.back),
    btn(t(lang, "btnNext"), CB.next),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function hourKeyboard(selected: number | undefined, lang: Lang): InlineKeyboardMarkup {
  const rows: InlineKeyboardButton[][] = [];
  let row: InlineKeyboardButton[] = [];
  for (let h = 0; h < 24; h++) {
    row.push(btn(`${String(h).padStart(2, "0")}:00${h === selected ? " ✓" : ""}`, `${CB.hourPick}:${h}`));
    if (row.length === 6) {
      rows.push(row);
      row = [];
    }
  }
  if (row.length) rows.push(row);
  rows.push([
    btn(t(lang, "btnBack"), CB.back),
    btn(t(lang, "btnNext"), CB.next),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function quarterKeyboard(selected: number | undefined, lang: Lang): InlineKeyboardMarkup {
  const ranges = [":00–:15", ":15–:30", ":30–:45", ":45–:00"];
  const rows: InlineKeyboardButton[][] = [];
  let row: InlineKeyboardButton[] = [];
  ranges.forEach((r, i) => {
    row.push(btn(`${r}${i === selected ? " ✓" : ""}`, `${CB.quarterPick}:${i}`));
    if (row.length === 2) {
      rows.push(row);
      row = [];
    }
  });
  if (row.length) rows.push(row);
  rows.push([
    btn(t(lang, "btnBack"), CB.back),
    btn(t(lang, "btnNext"), CB.next),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function callsignKeyboard(
  profile: Profile,
  lang: Lang,
  mode: "start" | "choose"
): InlineKeyboardMarkup | undefined {
  if (mode === "start") return undefined;
  const rows: InlineKeyboardButton[][] = [];
  if (profile.callsign) rows.push([btn(t(lang, "callsignUseDefault", { callsign: profile.callsign }), CB.callsignUse)]);
  rows.push([
    btn(t(lang, "callsignEnterNew"), CB.callsignNew),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function gridKeyboard(
  profile: Profile,
  lang: Lang,
  mode: "start" | "choose"
): InlineKeyboardMarkup | undefined {
  if (mode === "start") return undefined;
  const rows: InlineKeyboardButton[][] = [];
  if (profile.grid) rows.push([btn(t(lang, "gridUseDefault", { grid: profile.grid }), CB.gridUse)]);
  rows.push([
    btn(t(lang, "gridEnterNew"), CB.gridNew),
    btn(t(lang, "btnCancel"), CB.cancel),
  ]);
  return { inline_keyboard: rows };
}

export function confirmKeyboard(lang: Lang): InlineKeyboardMarkup {
  return {
    inline_keyboard: [
      [
        btn(t(lang, "btnSubmit"), CB.submit),
        btn(t(lang, "btnBack"), CB.back),
        btn(t(lang, "btnCancel"), CB.cancel),
      ],
    ],
  };
}

export function langKeyboard(): InlineKeyboardMarkup {
  return {
    inline_keyboard: [
      [btn("🇺🇸 English", `${CB.langPick}:en`), btn("🇷🇺 Русский", `${CB.langPick}:ru`)],
      [btn("🇨🇳 中文", `${CB.langPick}:zh`)],
      [btn("❌", CB.cancel)],
    ],
  };
}

export function mainMenuKeyboard(lang: Lang): InlineKeyboardMarkup {
  return {
    inline_keyboard: [
      [btn(t(lang, "btnMenuReport"), "menu:report")],
      [btn(t(lang, "btnMenuLang"), "menu:lang"), btn(t(lang, "btnMenuHelp"), "menu:help")],
    ],
  };
}

File: pybot/keyboards.py
# -*- coding: utf-8 -*-
"""Inline keyboard builders for the wizard steps."""

from telegram import InlineKeyboardButton, InlineKeyboardMarkup

from i18n import month_name, status_label, t

import base64


def b64url_encode(s):
    return base64.urlsafe_b64encode(s.encode("utf-8")).decode("ascii").rstrip("=")


def b64url_decode(s):
    pad = "=" * (4 - len(s) % 4)
    return base64.urlsafe_b64decode((s + pad).encode("ascii")).decode("utf-8")


def _b(text, data):
    return InlineKeyboardButton(text, callback_data=data)


def back_cancel_row(lang):
    return [_b(t(lang, "btnBack"), "bk"), _b(t(lang, "btnCancel"), "cc")]


def back_next_cancel_row(lang):
    return [_b(t(lang, "btnBack"), "bk"), _b(t(lang, "btnNext"), "nx"), _b(t(lang, "btnCancel"), "cc")]


def satellite_keyboard(sats, page, total_pages, lang):
    rows = []
    start = page * 8
    for s in sats[start : start + 8]:
        label = s.get("display_name") or s["name"]
        rows.append([_b(label, "sat:" + b64url_encode(s["name"]))])
    nav = []
    if page > 0:
        nav.append(_b("◀️", f"sp:{page - 1}"))
    if page < total_pages - 1:
        nav.append(_b("▶️", f"sp:{page + 1}"))
    if nav:
        rows.append(nav)
    rows.append([_b(t(lang, "btnSearch"), "ss"), _b(t(lang, "btnCancel"), "cc")])
    return InlineKeyboardMarkup(rows)


def status_keyboard(statuses, lang, selected=None):
    rows = []
    for s in statuses:
        label = status_label(lang, s["value"], s.get("label"))
        if s["value"] == selected:
            label += " ✓"
        rows.append([_b(label, "st:" + b64url_encode(s["value"]))])
    rows.append(back_cancel_row(lang))
    return InlineKeyboardMarkup(rows)


def date_keyboard(date_str, lang):
    return InlineKeyboardMarkup(
        [
            [_b(t(lang, "dateYes", {"date": date_str}), "dty"), _b(t(lang, "dateNo"), "dtn")],
            back_cancel_row(lang),
        ]
    )


def time_keyboard(time_str, lang):
    return InlineKeyboardMarkup(
        [
            [_b(t(lang, "timeYes", {"time": time_str}), "tmy"), _b(t(lang, "timeNo"), "tmn")],
            back_cancel_row(lang),
        ]
    )


def year_keyboard(current_year, page, total_pages, selected, lang):
    years = list(range(current_year, 2000, -1))
    rows = []
    start = page * 10
    row = []
    for y in years[start : start + 10]:
        label = f"{y} ✓" if y == selected else str(y)
        row.append(_b(label, f"yr:{y}"))
        if len(row) == 5:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    nav = []
    if page > 0:
        nav.append(_b("◀️", f"yp:{page - 1}"))
    if page < total_pages - 1:
        nav.append(_b("▶️", f"yp:{page + 1}"))
    if nav:
        rows.append(nav)
    rows.append(back_next_cancel_row(lang))
    return InlineKeyboardMarkup(rows)


def month_keyboard(selected, lang):
    rows = []
    row = []
    for m in range(1, 13):
        label = month_name(lang, m) + (" ✓" if m == selected else "")
        row.append(_b(label, f"mo:{m}"))
        if len(row) == 4:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    rows.append(back_next_cancel_row(lang))
    return InlineKeyboardMarkup(rows)


def day_keyboard(year, month, selected, lang):
    import calendar

    days = calendar.monthrange(year, month)[1]
    rows = []
    row = []
    for d in range(1, days + 1):
        label = f"{d} ✓" if d == selected else str(d)
        row.append(_b(label, f"dy:{d}"))
        if len(row) == 7:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    rows.append(back_next_cancel_row(lang))
    return InlineKeyboardMarkup(rows)


def hour_keyboard(selected, lang):
    rows = []
    row = []
    for h in range(24):
        label = f"{h:02d}:00 ✓" if h == selected else f"{h:02d}:00"
        row.append(_b(label, f"hr:{h}"))
        if len(row) == 6:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    rows.append(back_next_cancel_row(lang))
    return InlineKeyboardMarkup(rows)


def quarter_keyboard(selected, lang):
    ranges = [":00–:15", ":15–:30", ":30–:45", ":45–:00"]
    rows = []
    row = []
    for i, r in enumerate(ranges):
        label = r + (" ✓" if i == selected else "")
        row.append(_b(label, f"qr:{i}"))
        if len(row) == 2:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    rows.append(back_next_cancel_row(lang))
    return InlineKeyboardMarkup(rows)


def callsign_keyboard(profile, lang):
    if not profile.get("callsign"):
        return None
    rows = [[_b(t(lang, "callsignUseDefault", {"callsign": profile["callsign"]}), "csu")]]
    rows.append([_b(t(lang, "callsignEnterNew"), "csn"), _b(t(lang, "btnCancel"), "cc")])
    return InlineKeyboardMarkup(rows)


def grid_keyboard(profile, lang):
    if not profile.get("grid"):
        return None
    rows = [[_b(t(lang, "gridUseDefault", {"grid": profile["grid"]}), "gru")]]
    rows.append([_b(t(lang, "gridEnterNew"), "grn"), _b(t(lang, "btnCancel"), "cc")])
    return InlineKeyboardMarkup(rows)


def confirm_keyboard(lang):
    return InlineKeyboardMarkup(
        [
            [_b(t(lang, "btnSubmit"), "cfy"), _b(t(lang, "btnBack"), "bk"), _b(t(lang, "btnCancel"), "cc")],
        ]
    )


def lang_keyboard():
    return InlineKeyboardMarkup(
        [
            [_b("🇺🇸 English", "ln:en"), _b("🇷🇺 Русский", "ln:ru")],
            [_b("🇨🇳 中文", "ln:zh")],
            [_b("❌", "cc")],
        ]
    )


def main_menu_keyboard(lang):
    return InlineKeyboardMarkup(
        [
            [_b(t(lang, "btnMenuReport"), "menu:report")],
            [_b(t(lang, "btnMenuLang"), "menu:lang"), _b(t(lang, "btnMenuHelp"), "menu:help")],
        ]
    )